File: scripts/import-players.ts
import { PrismaClient, type Prisma } from "@prisma/client";
import axios from "axios";
import { parse } from "csv-parse";
import { Readable } from "stream";

const prisma = new PrismaClient({
  log: ["error"],
});

interface NFLPlayer {
  first_name: string;
  last_name: string;
  position: string | null;
  esb_id: string | null;
  gsis_id: string;
  display_name: string | null;
  rookie_year: string | null;
  college_conference: string | null;
  current_team_id: string | null;
  draft_club: string | null;
  draft_number: string | null;
  draft_round: string | null;
  entry_year: string | null;
  football_name: string | null;
  gsis_it_id: string | null;
  headshot: string | null;
  jersey_number: string | null;
  position_group: string | null;
  short_name: string | null;
  smart_id: string | null;
  status: string | null;
  status_description_abbr: string | null;
  status_short_description: string | null;
  team_abbr: string | null;
  uniform_number: string | null;
  height: string | null;
  weight: string | null;
  college_name: string | null;
  years_of_experience: string | null;
  birth_date: string | null;
  team_seq: string | null;
  suffix: string | null;
}

interface CSVRecord {
  first_name: string;
  last_name: string;
  position: string;
  esb_id: string;
  gsis_id: string;
  display_name: string;
  rookie_year: string;
  college_conference: string;
  current_team_id: string;
  draft_club: string;
  draft_number: string;
  draft_round: string;
  entry_year: string;
  football_name: string;
  gsis_it_id: string;
  headshot: string;
  jersey_number: string;
  position_group: string;
  short_name: string;
  smart_id: string;
  status: string;
  status_description_abbr: string;
  status_short_description: string;
  team_abbr: string;
  uniform_number: string;
  height: string;
  weight: string;
  college_name: string;
  years_of_experience: string;
  birth_date: string;
  team_seq: string;
  suffix: string;
}

async function downloadPlayerData(): Promise<string> {
  try {
    console.log("Downloading player data...");
    const url =
      "https://github.com/nflverse/nflverse-data/releases/download/players/players.csv";
    const response = await axios.get<string>(url, { responseType: "text" });
    return response.data;
  } catch (error) {
    if (axios.isAxiosError(error)) {
      throw new Error(`Failed to download player data: ${error.message}`);
    }
    throw new Error("Failed to download player data: Unknown error occurred");
  }
}

async function parseCSV(data: string): Promise<NFLPlayer[]> {
  return new Promise((resolve, reject) => {
    const records: NFLPlayer[] = [];
    const parser = parse({
      columns: true,
      skip_empty_lines: true,
      cast: true,
    });

    const stream = Readable.from(data);
    stream
      .pipe(parser)
      .on("data", (record: CSVRecord) => {
        // Convert all fields to strings or null
        const player: NFLPlayer = {
          first_name: String(record.first_name),
          last_name: String(record.last_name),
          position: record.position ? String(record.position) : null,
          esb_id: record.esb_id ? String(record.esb_id) : null,
          gsis_id: String(record.gsis_id),
          display_name: record.display_name
            ? String(record.display_name)
            : null,
          rookie_year: record.rookie_year ? String(record.rookie_year) : null,
          college_conference: record.college_conference
            ? String(record.college_conference)
            : null,
          current_team_id: record.current_team_id
            ? String(record.current_team_id)
            : null,
          draft_club: record.draft_club ? String(record.draft_club) : null,
          draft_number: record.draft_number
            ? String(record.draft_number)
            : null,
          draft_round: record.draft_round ? String(record.draft_round) : null,
          entry_year: record.entry_year ? String(record.entry_year) : null,
          football_name: record.football_name
            ? String(record.football_name)
            : null,
          gsis_it_id: record.gsis_it_id ? String(record.gsis_it_id) : null,
          headshot: record.headshot ? String(record.headshot) : null,
          jersey_number: record.jersey_number
            ? String(record.jersey_number)
            : null,
          position_group: record.position_group
            ? String(record.position_group)
            : null,
          short_name: record.short_name ? String(record.short_name) : null,
          smart_id: record.smart_id ? String(record.smart_id) : null,
          status: record.status ? String(record.status) : null,
          status_description_abbr: record.status_description_abbr
            ? String(record.status_description_abbr)
            : null,
          status_short_description: record.status_short_description
            ? String(record.status_short_description)
            : null,
          team_abbr: record.team_abbr ? String(record.team_abbr) : null,
          uniform_number: record.uniform_number
            ? String(record.uniform_number)
            : null,
          height: record.height ? String(record.height) : null,
          weight: record.weight ? String(record.weight) : null,
          college_name: record.college_name
            ? String(record.college_name)
            : null,
          years_of_experience: record.years_of_experience
            ? String(record.years_of_experience)
            : null,
          birth_date: record.birth_date ? String(record.birth_date) : null,
          team_seq: record.team_seq ? String(record.team_seq) : null,
          suffix: record.suffix ? String(record.suffix) : null,
        };
        records.push(player);
      })
      .on("end", () => resolve(records))
      .on("error", (error) => reject(error));
  });
}

async function importPlayers() {
  try {
    const data = await downloadPlayerData();
    const players = await parseCSV(data);

    console.log(`Processing ${players.length} players...`);
    let playersCreated = 0;

    // Process in chunks of 100
    for (let i = 0; i < players.length; i += 100) {
      const playerChunk = players.slice(i, i + 100);

      await prisma.$transaction(async (tx: Prisma.TransactionClient) => {
        for (const player of playerChunk) {
          try {
            const existingPlayer = await tx.player.findUnique({
              where: { gsis_id: player.gsis_id },
            });

            if (!existingPlayer) {
              await tx.player.create({
                data: {
                  first_name: player.first_name,
                  last_name: player.last_name,
                  position: player.position,
                  esb_id: player.esb_id,
                  gsis_id: player.gsis_id,
                  display_name: player.display_name,
                  rookie_year: player.rookie_year
                    ? parseInt(player.rookie_year)
                    : null,
                  college_conference: player.college_conference,
                  current_team_id: player.current_team_id,
                  draft_club: player.draft_club,
                  draft_number: player.draft_number
                    ? parseInt(player.draft_number)
                    : null,
                  draft_round: player.draft_round
                    ? parseInt(player.draft_round)
                    : null,
                  entry_year: player.entry_year
                    ? parseInt(player.entry_year)
                    : null,
                  football_name: player.football_name,
                  gsis_it_id: player.gsis_it_id,
                  headshot: player.headshot,
                  jersey_number: player.jersey_number,
                  position_group: player.position_group,
                  short_name: player.short_name,
                  smart_id: player.smart_id,
                  status: player.status,
                  status_description_abbr: player.status_description_abbr,
                  status_short_description: player.status_short_description,
                  team_abbr: player.team_abbr,
                  uniform_number: player.uniform_number,
                  height: player.height,
                  weight: player.weight ? parseInt(player.weight) : null,
                  college_name: player.college_name,
                  years_of_experience: player.years_of_experience
                    ? parseInt(player.years_of_experience)
                    : null,
                  birth_date: player.birth_date
                    ? new Date(player.birth_date)
                    : null,
                  team_seq: player.team_seq ? parseInt(player.team_seq) : null,
                  suffix: player.suffix,
                },
              });
              playersCreated++;
            }
          } catch (error) {
            console.error(`Error processing player ${player.gsis_id}:`, error);
            continue;
          }
        }
      });

      if (playersCreated % 100 === 0) {
        console.log(`Created ${playersCreated} players...`);
      }
    }

    console.log(
      `Completed importing players. Created ${playersCreated} players.`,
    );
  } catch (error) {
    console.error("Failed to import player data:", error);
    throw error;
  } finally {
    await prisma.$disconnect();
  }
}

async function main() {
  try {
    console.log("Starting player data import...");
    await importPlayers();
    console.log("Player data import completed");
  } catch (error) {
    console.error("Failed to import player data:", error);
    process.exit(1);
  }
}

void main().catch((error) => {
  console.error("Unhandled error:", error);
  process.exit(1);
});

File: scripts/import_players.py
import pandas as pd
import requests
import psycopg
from psycopg.rows import dict_row
from psycopg.copy import Copy
import logging
from datetime import datetime
from config import DATABASE_URL
import sys
from io import StringIO

# Set up logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

def download_player_data() -> pd.DataFrame:
    """Download player data from nflverse."""
    logger.info("Downloading player data...")
    url = "https://github.com/nflverse/nflverse-data/releases/download/players/players.csv"
    
    try:
        df = pd.read_csv(url, low_memory=False)
        logger.info(f"Downloaded {len(df)} players")
        return df
    except Exception as e:
        logger.error(f"Failed to download player data: {str(e)}")
        raise

def process_players(df: pd.DataFrame) -> pd.DataFrame:
    """Process players data for bulk insert."""
    # Initialize missing columns with None
    required_columns = {
        'commonFirstName': None,
        'nflId': None,
        'pfrId': None,
        'pffId': None,
        'otcId': None,
        'espnId': None,
        'ngsPositionGroup': None,
        'ngsPosition': None,
        'rookieSeason': None,
        'lastSeason': None,
        'latestTeam': None,
        'ngsStatus': None,
        'ngsStatusDescription': None,
        'pffPosition': None,
        'pffStatus': None,
        'draftYear': None,
        'draftRound': None,
        'draftPick': None,
        'draftTeam': None
    }
    
    for col, default_value in required_columns.items():
        if col not in df.columns:
            df[col] = default_value
    
    # Add timestamps
    now = datetime.utcnow()
    df['createdAt'] = now
    df['updatedAt'] = now
    
    # Rename columns to match our schema
    df = df.rename(columns={
        'gsis_id': 'gsisId',
        'display_name': 'displayName',
        'common_first_name': 'commonFirstName',
        'first_name': 'firstName',
        'last_name': 'lastName',
        'short_name': 'shortName',
        'football_name': 'footballName',
        'suffix': 'suffix',
        'esb_id': 'esbId',
        'nfl_id': 'nflId',
        'pfr_id': 'pfrId',
        'pff_id': 'pffId',
        'otc_id': 'otcId',
        'espn_id': 'espnId',
        'smart_id': 'smartId',
        'birth_date': 'birthDate',
        'position_group': 'positionGroup',
        'position': 'position',
        'ngs_position_group': 'ngsPositionGroup',
        'ngs_position': 'ngsPosition',
        'height': 'height',
        'weight': 'weight',
        'headshot': 'headshot',
        'college_name': 'collegeName',
        'college_conference': 'collegeConference',
        'jersey_number': 'jerseyNumber',
        'rookie_season': 'rookieSeason',
        'last_season': 'lastSeason',
        'latest_team': 'latestTeam',
        'status': 'status',
        'ngs_status': 'ngsStatus',
        'ngs_status_short_description': 'ngsStatusDescription',
        'years_of_experience': 'yearsOfExperience',
        'pff_position': 'pffPosition',
        'pff_status': 'pffStatus',
        'draft_year': 'draftYear',
        'draft_round': 'draftRound',
        'draft_pick': 'draftPick',
        'draft_team': 'draftTeam'
    })
    
    # Filter out rows where required fields are null
    df = df.dropna(subset=['gsisId', 'firstName', 'lastName'])
    
    # Set id to be the same as gsisId
    df['id'] = df['gsisId']
    
    # Drop any columns that don't match our schema
    schema_columns = [
        'id', 'position', 'headshot', 'status', 'height', 'weight',
        'suffix', 'createdAt', 'updatedAt', 'birthDate', 'collegeConference',
        'collegeName', 'commonFirstName', 'displayName', 'draftPick',
        'draftRound', 'draftTeam', 'draftYear', 'esbId', 'espnId',
        'firstName', 'footballName', 'gsisId', 'jerseyNumber', 'lastName',
        'lastSeason', 'latestTeam', 'nflId', 'ngsPosition', 'ngsPositionGroup',
        'ngsStatus', 'ngsStatusDescription', 'otcId', 'pffId', 'pffPosition',
        'pffStatus', 'pfrId', 'positionGroup', 'rookieSeason', 'shortName',
        'smartId', 'yearsOfExperience'
    ]
    df = df[schema_columns]
    
    # Convert numeric fields
    numeric_fields = {
        'rookieSeason': 'Int64',
        'lastSeason': 'Int64',
        'weight': 'Int64',
        'yearsOfExperience': 'Int64',
        'draftYear': 'Int64',
        'draftRound': 'Int64',
        'draftPick': 'Int64'
    }
    
    for field, dtype in numeric_fields.items():
        if field in df.columns:
            df[field] = pd.to_numeric(df[field], errors='coerce').astype(dtype)
    
    # Convert date fields
    df['birthDate'] = pd.to_datetime(df['birthDate'], errors='coerce')
    
    return df

def bulk_insert_dataframe(conn: psycopg.Connection, df: pd.DataFrame, table: str, conflict_fields: list):
    """Bulk insert a dataframe using COPY command."""
    # Create temporary table
    temp_table = f"temp_{table}"
    column_types = {
        'id': 'text',
        'position': 'text',
        'headshot': 'text',
        'status': 'text',
        'height': 'text',
        'weight': 'integer',
        'suffix': 'text',
        'createdAt': 'timestamp(3) without time zone',
        'updatedAt': 'timestamp(3) without time zone',
        'birthDate': 'timestamp(3) without time zone',
        'collegeConference': 'text',
        'collegeName': 'text',
        'commonFirstName': 'text',
        'displayName': 'text',
        'draftPick': 'integer',
        'draftRound': 'integer',
        'draftTeam': 'text',
        'draftYear': 'integer',
        'esbId': 'text',
        'espnId': 'text',
        'firstName': 'text',
        'footballName': 'text',
        'gsisId': 'text',
        'jerseyNumber': 'text',
        'lastName': 'text',
        'lastSeason': 'integer',
        'latestTeam': 'text',
        'nflId': 'text',
        'ngsPosition': 'text',
        'ngsPositionGroup': 'text',
        'ngsStatus': 'text',
        'ngsStatusDescription': 'text',
        'otcId': 'text',
        'pffId': 'text',
        'pffPosition': 'text',
        'pffStatus': 'text',
        'pfrId': 'text',
        'positionGroup': 'text',
        'rookieSeason': 'integer',
        'shortName': 'text',
        'smartId': 'text',
        'yearsOfExperience': 'integer'
    }
    
    columns = df.columns.tolist()
    create_temp_table = f"""
        CREATE TEMP TABLE {temp_table} (
            {', '.join(f'"{col}" {column_types.get(col, "text")}' for col in columns)}
        ) ON COMMIT DROP
    """
    
    with conn.cursor() as cur:
        # Create temp table
        logger.info(f"Creating temp table: {create_temp_table}")
        cur.execute(create_temp_table)
        
        # Copy data to temp table
        buffer = StringIO()
        # Replace NaN with None for proper NULL handling
        df = df.replace({pd.NA: None, pd.NaT: None})
        df.to_csv(buffer, index=False, header=False, na_rep='\\N')
        buffer.seek(0)
        
        copy_cmd = f"COPY {temp_table} FROM STDIN WITH CSV NULL '\\N'"
        logger.info(f"Copying data: {copy_cmd}")
        with cur.copy(copy_cmd) as copy:
            copy.write(buffer.getvalue())
        
        # Perform upsert from temp table
        conflict_cols = ', '.join(f'"{f}"' for f in conflict_fields)
        update_cols = [col for col in columns if col not in conflict_fields and col != 'updatedAt']
        update_stmt = ', '.join(f'"{col}" = EXCLUDED."{col}"' for col in update_cols)
        if update_stmt:
            update_stmt += ', '
        update_stmt += '"updatedAt" = CURRENT_TIMESTAMP'
        
        upsert_query = f"""
            INSERT INTO "{table}" ({', '.join(f'"{col}"' for col in columns)})
            SELECT {', '.join(f'"{col}"::{column_types.get(col, "text")}' for col in columns)}
            FROM {temp_table}
            ON CONFLICT ({conflict_cols})
            DO UPDATE SET {update_stmt}
        """
        logger.info(f"Upserting data: {upsert_query}")
        cur.execute(upsert_query)
        
        # Verify the data was inserted
        verify_query = f'SELECT COUNT(*) FROM "{table}"'
        cur.execute(verify_query)
        count = cur.fetchone()[0]
        logger.info(f"Verified {count} rows in {table} table")

def import_players():
    """Import player data."""
    try:
        # Download and process data
        data = download_player_data()
        players = process_players(data)

        logger.info(f"Processing {len(players)} players...")
        logger.info("Importing players...")

        # Connect to database
        conn = psycopg.connect(DATABASE_URL)
        
        try:
            # Start transaction
            with conn.transaction():
                # Process players data
                df = pd.DataFrame(players)
                df = process_players(df)
                
                # Bulk insert players
                bulk_insert_dataframe(conn, df, "Player", ["id"])
                
                # Verify the data was inserted
                with conn.cursor() as cur:
                    cur.execute('SELECT COUNT(*) FROM "Player"')
                    count = cur.fetchone()[0]
                    logger.info(f"Inserted {count} players")
                
                logger.info("Successfully imported players")
            
        except Exception as e:
            logger.error(f"Error during import: {str(e)}")
            raise
        finally:
            conn.close()
            
    except Exception as e:
        logger.error(f"Failed to import player data: {str(e)}")
        raise

def main():
    """Main function to import player data."""
    try:
        logger.info("Starting player data import...")
        import_players()
        logger.info("Player data import completed")
    except Exception as e:
        logger.error(f"Failed to import player data: {str(e)}")
        sys.exit(1)

if __name__ == "__main__":
    main() 